adTechModel.py

```diff
@@ -0,0 +1,329 @@
+import yaml
+from yaml.loader import SafeLoader
+import argparse
+
+# Parse command line to get company file
+parser = argparse.ArgumentParser(description="Compute emissions for an ad tech company")
+parser.add_argument(
+    "-d",
+    "--defaultsFile",
+    default="defaults.yaml",
+    help="Set the defaults file to use (overrides defaults.yaml)",
+)
+parser.add_argument("-v", "--verbose", action="store_true", help="Show derivation of output")
+parser.add_argument(
+    "-p", "--partners", const=1, default=0, type=int, nargs="?", help="Simulate distribution partners"
+)
+parser.add_argument("companyFile", nargs=1, help="The company file to parse in YAML format")
+args = parser.parse_args()
+
+if args.verbose:
+
+    def verboseprint(*args, **kwargs) -> None:
+        print(*args, **kwargs)
+
+else:
+    verboseprint = lambda *a, **k: None  # do-nothing function
+
+# Load defaults
+defaultsStream = open(args.defaultsFile, "r")
+defaultsDocument = yaml.load(defaultsStream, Loader=SafeLoader)
+
+# Load facts about the company
+facts: dict[str, float] = {}
+stream = open(args.companyFile[0], "r")
+documents = list(yaml.load_all(stream, Loader=SafeLoader))
+template: str
+document = documents[0]  # shouldn't have multiple
+if "company" not in document:
+    raise Exception("No 'company' field found in company file")
+if "products" not in document["company"]:
+    raise Exception("No 'products' field found in company file")
+if "sources" not in document["company"]:
+    raise Exception("No 'sources' field found in company file")
+for source in document["company"]["sources"]:
+    for fact in source["source"]["facts"]:
+        keys = [key for key in fact["fact"] if key != "reference" and key != "comment"]
+        for key in keys:
+            facts[key] = fact["fact"][key]
+
+distributionPartners = []
+for i in range(args.partners):
+
+    distributionPartners.append(
+        {
+            "name": "dummy",
+            "identifier": "dummy.org",
+            "primaryEmissionsGPerBidRequest": 0.001,
```

Comment:
Should this be `0.0001` ?

Reply:
doesn't really matter since it's dummy data but prob